### ISSUE TITLE
unbalanced data distribution; early chunks deletion

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#502]
I had an interesting incident recently: 3TB HDD was physically disconnected (pulled out) by mistake on working chunkserver. Of course LizardFS quickly marked HDD as "damaged" and started replication. Somewhat 16...20 hours later I've noticed the problem by looking at LizardFS CGI interface. By that time most of the data (~90%) from unavailable HDD was replicated. I reconnected HDD (re-mounted file system etc.) and reloaded chunkserver at which time I've noticed something strange: chunkserver was scanning available chunks very slowly. As I've realised later chunkserver was deleting chunks from this HDD even during scanning (or very early) because shortly after HDD was fully scanned it ended up only 10% full. It was above 80% full HDD when I plugged it. Massive deletion of chunks manifested as spike on corresponding "number of chunks deletions per minute" graph. In the end reconnected HDD was almost empty so LizardFS started replicating data again and now data flows back to reconnected HDD.
What should have happened is extra overgoal chunks should have been deleted not only from reconnected HDD but from all over the cluster to achieve balanced distribution of data and to prevent needless rebalancing...

May be related to #491 but the actual problem here is not as much with data distribution but with (incorrect) massive early deletion of overgoal chunks from re-added storage.

**Comment by psarna** [comment]
Was it possible that chunks from this HDD were already in outdated version? If so, then they were all valid candidates for removal. Anyway, I'll look into it more closely on Monday, if there is an issue with master server picking chunks for deletion prematurely or in an unbalanced way, it will be fixed.

**Comment by 4Dolio** [comment]
What you described is normal. I have leveraged this behaviour when i manually redistribute chunks to a new disk in a chink server. You should have onlined the recovered disk first by itself, and then the others (via reload) to get the desired results. 

for example. I manually transfer 1/20th of chunks from 20 existing disks to 1 new disk. Then startbwith the 1 new disk only. Then hot relaod to online the 20 original disks which when scanned will purge the chunks which were duplicated to the new disk from themselves...  I even devosed a script which takes a list of source and target disks and a ratio to transfer and an option to move them (while offline) or copy them (while online). Then they are carefully brought back into service in a specific order.

**Comment by 4Dolio** [comment]
It comes down to this. If multiple copies of a chunk exist on a single chunk server. The 2nd copy is purged during startup/reload scan of chunks.  You can control which are purged by controlling the order the disks get scanned by enabling specific disks with hdd.cfg changes and service reloads.

**Comment by onlyjob** [comment]
> Was it possible that chunks from this HDD were already in outdated version?

No, not possible... It could be that around 5% of chunks were re-written while HDD was down but not massive 90% of 3TB during hours of downtime...

Thanks. So the problem was that I've re-added disk by reloading chunkserver when all other HDDs were already scanned/operational? I suppose it is somewhat unexpected to delete data from HDD during scanning... Even if proper distribution could be achieved by restarting chunkserver (instead of reloading its configuration with re-added HDD) I think master should probably orchestrate chunks deletion as usual... I'm not sure how to handle multiple copies of chunks on the same chunkserver but deleting 'em only from newly added HDD feels wrong...

**Comment by 4Dolio** [comment]
Absolutle reasonable to be caught off guard by this unexpected behaviour. While i dont think the behavour needs to change.  I do think it needs to be well explained in documentation...

**Comment by onlyjob** [comment]
So chunkservers delete chunks during scanning because they can't handle multiple copies of chunks, right? That make sense...

**Comment by 4Dolio** [comment]
@onlyjob, You needed to have:
* chunkserver: stop, start with only temporarially lost disk, enable all remaining disks and reload.
* it is still possible for you to redistribute the chunks, but it is not as simple as:
** lizardfs chunkrebalance servername

**Comment by 4Dolio** [comment]
Add an option: RmDups {live,scan} on chunkservers service. Both are useful.
Worth pondering if live should be default. If you bring a disk online with valid paths to chunks assume you want those to be prefered over live duplicate chunks.

**Comment by 4Dolio** [comment]
https://github.com/lizardfs/lizardfs/issues/443#issuecomment-240623908 is related I believe
